From ae6f2ccda6b5c349d237b364bf7d413bd2094f39 Mon Sep 17 00:00:00 2001
From: Kent Fredric <kentnl@gentoo.org>
Date: Sun, 16 May 2021 15:00:02 +0200
Subject: [PATCH 30/30] Add support for -Dgentoolibdirs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Which just adds the libdirs verbatim in the right place without
having perl molest it in the process.

Ported to 5.36 by Andreas K. Hüttel <dilfridge@gentoo.org>

Signed-off-by: Andreas K. Hüttel <dilfridge@gentoo.org>
---
 Configure                      | 26 ++++++++++++++++++++++++++
 Cross/config.sh-arm-linux      |  2 ++
 Cross/config.sh-arm-linux-n770 |  2 ++
 Porting/config.sh              |  2 ++
 config_h.SH                    |  5 +++++
 configure.com                  |  2 ++
 perl.c                         |  1 +
 perl_inc_macro.h               |  8 ++++++++
 plan9/config_sh.sample         |  2 ++
 uconfig.h                      | 11 ++++++++---
 uconfig.sh                     |  2 ++
 uconfig64.sh                   |  2 ++
 win32/config.gc                |  2 ++
 win32/config.vc                |  2 ++
 15 files changed, 68 insertions(+), 3 deletions(-)

diff -ruN perl-5.36.0-RC2.orig/config_h.SH perl-5.36.0-RC2/config_h.SH
--- perl-5.36.0-RC2.orig/config_h.SH	2021-11-02 17:23:36.000000000 +0100
+++ perl-5.36.0-RC2/config_h.SH	2022-05-21 16:32:29.355714359 +0200
@@ -1308,6 +1308,11 @@
  */
 #$d_perl_otherlibdirs PERL_OTHERLIBDIRS "$otherlibdirs"		/**/
 
+/* GENTOO_LIBDIRS:
+ * Like PERL_OTHERLIBDIRS, but doesn't stuff ARCH dirs in when not wanted
+ */
+#$d_gentoolibdirs GENTOO_LIBDIRS "$gentoolibdirs"		/**/
+
 /* PRIVLIB:
  *	This symbol contains the name of the private library for this package.
  *	The library is private in the sense that it needn't be in anyone's
diff -ruN perl-5.36.0-RC2.orig/Configure perl-5.36.0-RC2/Configure
--- perl-5.36.0-RC2.orig/Configure	2022-05-21 16:31:42.426057786 +0200
+++ perl-5.36.0-RC2/Configure	2022-05-21 16:32:23.912638208 +0200
@@ -1169,6 +1169,8 @@
 ranlib=''
 d_perl_otherlibdirs=''
 otherlibdirs=''
+gentoolibdirs=''
+d_gentoolibdirs=''
 package=''
 spackage=''
 pager=''
@@ -8056,6 +8058,28 @@
 set d_perl_otherlibdirs
 eval $setvar
 
+case "$gentoolibdirs" in
+''|' ') dflt='none' ;;
+*) dflt="$gentoolibdirs" ;;
+esac
+$cat <<EOM
+Enter a colon-seperated list of explicit gentoo paths to stuff in @INC
+unmolested, or enter 'none' for no extra paths
+
+EOM
+rp='Colon-seperated list of gentoo-specific perl library search dirs?'
+. ./myread
+case "$ans" in
+' '|''|none) gentoolibdirs=' ';;
+*) gentoolibdirs="$ans" ;;
+esac
+case "$gentoolibdirs" in
+' ') val=$undef ;;
+*) val=$define ;;
+esac
+set d_gentoolibdirs
+eval $setvar
+
 : DTrace support
 dflt_dtrace='/usr/sbin/dtrace'
 $test -x /usr/bin/dtrace && dflt_dtrace='/usr/bin/dtrace'
@@ -24770,6 +24794,7 @@
 d_pathconf='$d_pathconf'
 d_pause='$d_pause'
 d_perl_otherlibdirs='$d_perl_otherlibdirs'
+d_gentoolibdirs='$d_gentoolibdirs'
 d_phostname='$d_phostname'
 d_pipe2='$d_pipe2'
 d_pipe='$d_pipe'
@@ -25307,6 +25332,7 @@
 osname='$osname'
 osvers='$osvers'
 otherlibdirs='$otherlibdirs'
+gentoolibdirs='$gentoolibdirs'
 package='$package'
 pager='$pager'
 passcat='$passcat'
diff -ruN perl-5.36.0-RC2.orig/configure.com perl-5.36.0-RC2/configure.com
--- perl-5.36.0-RC2.orig/configure.com	2022-05-20 02:01:22.000000000 +0200
+++ perl-5.36.0-RC2/configure.com	2022-05-21 16:32:29.357714387 +0200
@@ -6683,6 +6683,7 @@
 $ WC "full_csh='" + " '"
 $ WC "full_sed='_NLA0:'"
 $ WC "gccversion='" + gccversion + "'"
+$ WC "gentoolibdirs='" + "'"
 $ WC "gidformat='lu'"
 $ WC "gidsign='1'"
 $ WC "gidsize='4'"
@@ -7084,6 +7085,7 @@
 $ WC "d_endservent_r='undef'"
 $ WC "d_freelocale='undef'"
 $ WC "d_gai_strerror='define'"
+$ WC "d_gentoolibdirs='undef'"
 $ WC "d_getgrent_r='undef'"
 $ WC "d_getgrgid_r='" + d_getgrgid_r + "'"
 $ WC "d_getgrnam_r='" + d_getgrnam_r + "'"
diff -ruN perl-5.36.0-RC2.orig/Cross/config.sh-arm-linux perl-5.36.0-RC2/Cross/config.sh-arm-linux
--- perl-5.36.0-RC2.orig/Cross/config.sh-arm-linux	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/Cross/config.sh-arm-linux	2022-05-21 16:32:23.912638208 +0200
@@ -248,6 +248,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -695,6 +696,7 @@
 gccansipedantic=''
 gccosandvers=''
 gccversion='2.95.3 20010125 (prerelease)'
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/Cross/config.sh-arm-linux-n770 perl-5.36.0-RC2/Cross/config.sh-arm-linux-n770
--- perl-5.36.0-RC2.orig/Cross/config.sh-arm-linux-n770	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/Cross/config.sh-arm-linux-n770	2022-05-21 16:32:23.912638208 +0200
@@ -247,6 +247,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -693,6 +694,7 @@
 gccansipedantic=''
 gccosandvers=''
 gccversion='2.95.3 20010125 (prerelease)'
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/perl.c perl-5.36.0-RC2/perl.c
--- perl-5.36.0-RC2.orig/perl.c	2022-05-21 16:31:38.475002504 +0200
+++ perl-5.36.0-RC2/perl.c	2022-05-21 16:32:29.357714387 +0200
@@ -4762,6 +4762,7 @@
     INCPUSH_ARCHLIB_EXP
     INCPUSH_PRIVLIB_EXP
     INCPUSH_PERL_OTHERLIBDIRS
+    INCPUSH_GENTOO_LIBDIRS
     INCPUSH_PERL5LIB
     INCPUSH_APPLLIB_OLD_EXP
     INCPUSH_PERL_OTHERLIBDIRS_ARCHONLY
diff -ruN perl-5.36.0-RC2.orig/perl_inc_macro.h perl-5.36.0-RC2/perl_inc_macro.h
--- perl-5.36.0-RC2.orig/perl_inc_macro.h	2021-10-15 15:28:26.000000000 +0200
+++ perl-5.36.0-RC2/perl_inc_macro.h	2022-05-21 16:32:29.357714387 +0200
@@ -141,6 +141,14 @@
                       INCPUSH_ADD_OLD_VERS|INCPUSH_CAN_RELOCATE);
 #endif
 
+#ifdef GENTOO_LIBDIRS
+# define INCPUSH_GENTOO_LIBDIRS S_incpush_use_sep(aTHX_ STR_WITH_LEN(GENTOO_LIBDIRS), \
+    INCPUSH_ADD_OLD_VERS|INCPUSH_CAN_RELOCATE);
+#endif
+#ifndef INCPUSH_GENTOO_LIBDIRS
+# define INCPUSH_GENTOO_LIBDIRS
+#endif
+
 #ifdef PERL_OTHERLIBDIRS
 #	define INCPUSH_PERL_OTHERLIBDIRS_ARCHONLY  S_incpush_use_sep(aTHX_ STR_WITH_LEN(PERL_OTHERLIBDIRS), \
                       INCPUSH_ADD_OLD_VERS|INCPUSH_ADD_ARCHONLY_SUB_DIRS|INCPUSH_CAN_RELOCATE);
diff -ruN perl-5.36.0-RC2.orig/plan9/config_sh.sample perl-5.36.0-RC2/plan9/config_sh.sample
--- perl-5.36.0-RC2.orig/plan9/config_sh.sample	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/plan9/config_sh.sample	2022-05-21 16:32:29.357714387 +0200
@@ -248,6 +248,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -693,6 +694,7 @@
 full_sed='/bin/sed'
 gccosandvers=''
 gccversion=''
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/Porting/config.sh perl-5.36.0-RC2/Porting/config.sh
--- perl-5.36.0-RC2.orig/Porting/config.sh	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/Porting/config.sh	2022-05-21 16:32:29.353714330 +0200
@@ -264,6 +264,7 @@
 d_gai_strerror='define'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='define'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -719,6 +720,7 @@
 gccansipedantic=''
 gccosandvers=''
 gccversion='10.2.1 20200805 [revision dda1e9d08434def88ed86557d08b23251332c5aa]'
+gentoolibdirs=''
 getgrent_r_proto='REENTRANT_PROTO_I_SBWR'
 getgrgid_r_proto='REENTRANT_PROTO_I_TSBWR'
 getgrnam_r_proto='REENTRANT_PROTO_I_CSBWR'
diff -ruN perl-5.36.0-RC2.orig/uconfig64.sh perl-5.36.0-RC2/uconfig64.sh
--- perl-5.36.0-RC2.orig/uconfig64.sh	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/uconfig64.sh	2022-05-21 16:32:29.358714401 +0200
@@ -187,6 +187,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='undef'
 d_getenv_preserves_other_thread='define'
@@ -615,6 +616,7 @@
 freetype=void
 full_csh=''
 full_sed=''
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/uconfig.h perl-5.36.0-RC2/uconfig.h
--- perl-5.36.0-RC2.orig/uconfig.h	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/uconfig.h	2022-05-21 16:35:49.022552613 +0200
@@ -1273,6 +1273,11 @@
  */
 /*#define PERL_OTHERLIBDIRS " "		/ **/
 
+/* GENTOO_LIBDIRS:
+ * Like PERL_OTHERLIBDIRS, but doesn't stuff ARCH dirs in when not wanted
+ */
+/*#define GENTOO_LIBDIRS ""		/ **/
+
 /* PRIVLIB:
  *	This symbol contains the name of the private library for this package.
  *	The library is private in the sense that it needn't be in anyone's
@@ -4200,7 +4205,7 @@
  */
 /*#define	HAS_STDIO_STREAM_ARRAY	/ **/
 #ifdef HAS_STDIO_STREAM_ARRAY
-#define STDIO_STREAM_ARRAY	
+#define STDIO_STREAM_ARRAY
 #endif
 
 /* GMTIME_MAX:
@@ -5340,6 +5345,6 @@
 #endif
 
 /* Generated from:
- * 87e5998978daf803d19866c43bca24d7c01dc74119650db16f8d18d83f355da9 config_h.SH
- * 1a5fe19cbcfd68ba70230580fd344189b5c78c11b2285efd5976366e51b3257e uconfig.sh
+ * bcfe9a9ebffbd54881138eae378b44d8256b640852de810009c910b3b09b0d47 config_h.SH
+ * 9c21f6e21ff6dba96bed87df4ae0e26c7f310dd006c004578d58287dca9152fc uconfig.sh
  * ex: set ro: */
diff -ruN perl-5.36.0-RC2.orig/uconfig.sh perl-5.36.0-RC2/uconfig.sh
--- perl-5.36.0-RC2.orig/uconfig.sh	2022-05-20 19:59:11.000000000 +0200
+++ perl-5.36.0-RC2/uconfig.sh	2022-05-21 16:32:29.358714401 +0200
@@ -187,6 +187,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='undef'
 d_getenv_preserves_other_thread='define'
@@ -615,6 +616,7 @@
 freetype=void
 full_csh=''
 full_sed=''
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/win32/config.gc perl-5.36.0-RC2/win32/config.gc
--- perl-5.36.0-RC2.orig/win32/config.gc	2022-05-20 02:01:22.000000000 +0200
+++ perl-5.36.0-RC2/win32/config.gc	2022-05-21 16:32:29.358714401 +0200
@@ -235,6 +235,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -686,6 +687,7 @@
 gccansipedantic=''
 gccosandvers=''
 gccversion=''
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
diff -ruN perl-5.36.0-RC2.orig/win32/config.vc perl-5.36.0-RC2/win32/config.vc
--- perl-5.36.0-RC2.orig/win32/config.vc	2022-05-20 02:01:22.000000000 +0200
+++ perl-5.36.0-RC2/win32/config.vc	2022-05-21 16:32:29.358714401 +0200
@@ -235,6 +235,7 @@
 d_gai_strerror='undef'
 d_gdbm_ndbm_h_uses_prototypes='undef'
 d_gdbmndbm_h_uses_prototypes='undef'
+d_gentoolibdirs='undef'
 d_getaddrinfo='undef'
 d_getcwd='define'
 d_getenv_preserves_other_thread='define'
@@ -685,6 +686,7 @@
 gccansipedantic=''
 gccosandvers=''
 gccversion=''
+gentoolibdirs=''
 getgrent_r_proto='0'
 getgrgid_r_proto='0'
 getgrnam_r_proto='0'
